ONSdigital/rm-collection-exercise-service
# Fix race condition due to state transitions out of sequence

# Motivation and Context
Acceptance tests are hanging because collection exercises are getting stuck in the wrong state and will never transition, because the sample units are all delivered before the state has been successfully persisted as EXECUTION_STARTED, so it can't move to EXECUTION_COMPLETE when the final sample unit is received.

This is all slightly liable to have very rare problems, when the sample service is temporarily unavailable, because of the _really bad_ mixture of synchronous and asynchronous and the lack of a distributed transaction when we do that. However, short of re-architecting the whole system, this will work 99.999% of the time and will be easy enough to fix in production if there's a catastrophic failure of multiple parts of the system.

# What has changed
Renamed the class which is responsible for preparing the CE to receive sample units. Updated it to do the state transition too, and commit the transaction BEFORE requesting the sample service to send over the sample units.

# How to test?
Upload lots of sample units and check that the CE goes to EXECUTION_COMPLETE without a problem. Running the acceptance test several times *might* reproduce it, but it was an intermittent bug we don't very often see on our local dev machines. If this fixes it, we won't see the intermittent bug anymore.

# Links
Trello: https://trello.com/c/KmhwKbXE/410-bug-state-transition-race-condition-when-delivering-samples-to-collex
